py-development.Rmd

```diff
@@ -0,0 +1,520 @@
+
+Up until now, our function calls have included enough arguments to assign one
+to each of the function parameters. This can get tedious in functions with many
+parameters and it is often beneficial to define default values for several of
+the function parameters. As an example, we can modify `sum_two_numbers()` to
+optionally return the two input arguments.
+
+```{python}
+def sum_two_numbers(num1, num2, return_input=False):
+    if return_input:
+        return num1, num2, num1 + num2
+    else:
+        return num1 + num2
+```
+
+By default the function will work just as previously.
+
+```{python}
+sum_two_numbers(2, 5)
+```
+
+But we now also have the option to return the input numbers.
+
+```{python}
+sum_two_numbers(2, 5, return_input=True)
+```
+
+Since the arguments are given in the same order as in the function definition,
+we could have left out the `return_input=` part and just written `True` in the
+third position. If we want to supply the arguments out of order, we need to
+specify the parameter name of each argument.
+
+#### Function documentation {py-function-documentation}
+
+Functions might appear self-explanatory when they are being written, but it is
+essential that there is proper documentation describing what the function does
+and what types of arguments should be in the input. This helps other people who
+are reading your code and also your future self that will be reusing these
+functions.
+
+In Python, a function is documented in its docstring, which is a multiline
+Python string immediately following the function definition. It is surrounded
+by triple quotes (either single or double) and can look like this.
+
+```{python}
+def sum_two_numbers(num1, num2):
+    '''
+    Add two numbers
+
+    Args:
+        num1 (int, float): The first number to be added.
+        num2 (int, float): The second number to be added.
+
+    Returns:
+        int, float: Sum of the two numbers.
+    '''
+
+    return num1 + num2
+```
+
+The above style is commonly referred to as the Google docstring convention. As
+functions get longer it can be beneficial to adapt the more elaborate `numpy`
+docstring convention used in many scientific Python packages:
+
+```{python}
+def sum_two_numbers(num1, num2):
+    '''
+    Add two numbers
+
+    Parameters
+    ----------
+    num1: int, float
+        The first number to be added.
+    num2: int, float
+        The second number to be added.
+
+    Returns
+    -------
+    int, float
+        Sum of the two numbers.
+    '''
+
+    return num1 + num2
+```
+
+We recommend using the `numpy` docstring format since this is easy to read
+complex function with many arguments and heavily used by the data science
+Python packages. Docstring are what show up in the function help messages, so
+it is important that these are well-written and helpful for the user.
+
+```{python}
+help(sum_two_numbers)
+```
+
+### Applying what we have learnt {py-applying-what-we-have-learnt}
```

Comment:
It would be great if they have written some small script previously in the course that we could modularize into functions here.